linode/manager
# upcoming: [M3-8020] – Add "Disk Encryption" section to Linode Rebuild modal

## Description 📝
Add "Disk Encryption" section to Linode Rebuild modal.

> [!NOTE]
> I came across an existing edge case where it is possible that a field absent from the Rebuild dialog can still cause the Rebuild request to fail silently. See M3-8237.

## Changes  🔄
- Add `isLKELinode`, `linodeIsInDistributedRegion`, and `isInRebuildFlow` props to `AccessPanel.tsx`
- Several props added to `RebuildFromImage` & `RebuildFromStackScript`

## Target release date 🗓️
6/24

## Preview 📷
<details>
<summary>Screenshots</summary>

**Standard Linode in a region that supports LDE**:

![Screenshot 2024-06-13 at 11 50 23 AM](https://github.com/linode/manager/assets/114682940/3347f2d5-3edb-463b-ae7d-6f9b45e4e867)

**Linode in Distributed region**:

![Screenshot 2024-06-13 at 11 27 05 AM](https://github.com/linode/manager/assets/114682940/3aaec5d0-3e44-458e-a6d5-dbc2cd8a8c8a)

**LKE Linode**:
![Screenshot 2024-06-13 at 11 49 26 AM](https://github.com/linode/manager/assets/114682940/86707b69-033f-4e4f-a200-e692005b2ea7)

</details>

## How to test 🧪

### Prerequisites
1. Have the `linode_disk_encryption` tag on your account
2. Point at the alpha/dev environment
3. Have the Linode Disk Encryption (LDE) flag in our dev tool toggled on

### Verification steps
Without the tag and/or with the LDE flag in our dev tool toggled off, you should not see any LDE-related things in the UI. Otherwise,

- for a non-LKE linode in Newark, the "Encrypt Disk" checkbox in the Rebuild modal should be enabled and you should be able to successfully rebuild it
- for an LKE linode, the "Encrypt Disk" checkbox in the Rebuild modal should be disabled with:
  - section description copy reading: `Secure this Linode using data at rest encryption. The disk encryption setting for Linodes added to a node pool will not be changed after rebuild.`
  - tooltip text reading: `The Encrypt Disk setting cannot be changed for a Linode attached to a node pool.`
- for a linode in a Distributed region,  the "Encrypt Disk" checkbox in the Rebuild modal should be disabled with:
  - section description copy reading: `Distributed Compute Instances are secured using disk encryption.`
  -  tooltip text reading: `The Encrypt Disk setting cannot be changed for distributed instances.`

The last two cases can be easily tested by adjusting this code block as needed and turning the MSW on:

https://github.com/linode/manager/blob/0e2d367d192f9a28b92b5b37ebd7f566c345ef5c/packages/manager/src/mocks/serverHandlers.ts#L768-L778

```typescript
      linodeFactory.build({
        backups: { enabled: false },
        id,
        label: 'Gecko Edge Test',
        lke_cluster_id: null, // change to a number to see the LKE Linode state
      })
```

Things should work when rebuilding from an image, a Community StackScript, and an Account StackScript.

Rebuilding an encrypted linode --> the "Encrypt Disk" checkbox should be checked already in the modal to reflect the current configuration

There should be no adverse effects observed if you switch to the prod environment and turn the feature flag off.

## As an Author I have considered 🤔
